include/deal.II/matrix_free/fe_remote_evaluation.h

```diff
@@ -999,6 +1073,373 @@ FERemoteEvaluationCommunicator<dim>::CopyInstructions::copy_data_entries(
+    const std::vector<
+      std::pair<types::boundary_id, std::function<std::vector<bool>()>>>
+                      &non_matching_faces_marked_vertices,
+    const unsigned int quad_no,
+    const unsigned int dof_no,
+    const double       tolerance)
+  {
+    const auto &dof_handler = matrix_free.get_dof_handler(dof_no);
+    const auto &tria        = dof_handler.get_triangulation();
+    const auto &mapping     = *matrix_free.get_mapping_info().mapping;
+
+    // Communication objects know about the communication pattern. I.e.,
+    // they know about the cells and quadrature points that have to be
+    // evaluated at remote faces. This information is given via
+    // RemotePointEvaluation. Additionally, the communication objects
+    // have to be able to match the quadrature points of the remote
+    // points (that provide exterior information) to the quadrature points
+    // defined at the interior cell. In case of point-to-point interpolation
+    // a vector of pairs with face batch Ids and the number of faces in the
+    // batch is needed. @c FERemoteCommunicationObjectEntityBatches
+    // is a container to store this information.
+    //
+    // We need multiple communication objects (one for each non-matching face
+    // ID).
+    std::vector<FERemoteCommunicationObjectEntityBatches<dim>> comm_objects;
+
+    // Additionally to the communication objects we need a vector
+    // that stores quadrature rules for every face batch.
+    // The quadrature can be empty in case of non non-matching faces,
+    // i.e. boundary faces. Internally this information is needed to correctly
+    // access values over multiple communication objects.
+    std::vector<Quadrature<dim>> global_quadrature_vector(
+      matrix_free.n_boundary_face_batches());
+
+    // Get the range of face batches we have to look at during construction of
+    // the communication objects. We only have to look at boundary faces.
+    const auto face_batch_range =
+      std::make_pair(matrix_free.n_inner_face_batches(),
+                     matrix_free.n_inner_face_batches() +
+                       matrix_free.n_boundary_face_batches());
+
+    // Iterate over all non-matching face IDs.
+    for (const auto &[nm_face, marked_vertices] :
+         non_matching_faces_marked_vertices)
+      {
+        // Construct the communication object for every face ID:
+        // 1) RemotePointEvaluation with user specified function for marked
+        // vertices.
+        auto rpe = std::make_shared<Utilities::MPI::RemotePointEvaluation<dim>>(
+          tolerance, false, 0, marked_vertices);
+
+        // 2) Face batch IDs and number of faces in batch.
+        std::vector<std::pair<unsigned int, unsigned int>>
+          face_batch_id_n_faces;
+
+        // Points that are searched by rpe.
+        std::vector<Point<dim>> points;
+
+        // Temporarily setup FEFaceEvaluation to access the quadrature points at
```

Comment:
```suggestion
        // Temporarily set up FEFaceEvaluation to access the quadrature points at
```
(you may want to look at all appearances, I'm going to stop highlighting them now.)